src/public/services/BetaService.ts

```diff
@@ -0,0 +1,77 @@
+import { get } from 'lodash'
+
+import { IFieldSchema, IForm } from '../../types'
+
+type BetaFeature = {
+  flag: string
+  matches: (field: IFieldSchema) => boolean
+}
+
+type BetaFeaturesDictionary = { [featureName: string]: BetaFeature }
+
+type HasBetaFlags = {
+  betaFlags?: {
+    [betaFlagName: string]: boolean
+  }
+}
+
+const BETA_FEATURES_FIELD: BetaFeaturesDictionary = {
+  // This is an example of how to add fields to this object
+  // featureName: {
+  //   flag: 'betaFlagName',
+  //   matches: (field) =>
+  //     field.fieldType === 'mobile' &&
+  //     (field.isVerifiable === true),
+  // },
+}
+
+const getBetaFeaturesForFields = (
+  formFields: IFieldSchema[] | undefined,
+  betaFeaturesField: BetaFeaturesDictionary,
+): string[] => {
+  const betaFeatures = new Set<string>()
+  if (formFields) {
+    for (const field of formFields) {
+      for (const [name, feature] of Object.entries(betaFeaturesField)) {
+        if (feature.matches(field)) {
+          betaFeatures.add(name)
+        }
+      }
+    }
+  }
+  return Array.from(betaFeatures)
+}
+
+export const getMissingFieldPermissions = (
+  user: HasBetaFlags,
+  form: IForm,
+  betaFeaturesField = BETA_FEATURES_FIELD,
+): string[] => {
+  const betaFeatures = getBetaFeaturesForFields(
+    form.form_fields,
+    betaFeaturesField,
+  )
+  return betaFeatures.filter((name) => {
+    const flag = get(betaFeaturesField, [name, 'flag'])
+    return flag && !get(user, ['betaFlags', flag], false)
+  })
+}
+
+export const isBetaField = (
+  fieldType: string,
+  betaFeaturesField = BETA_FEATURES_FIELD,
+): BetaFeature | undefined => {
+  return betaFeaturesField[fieldType]
+}
+
+export const userHasAccessToFieldType = (
+  user: HasBetaFlags,
```

Comment:
does the frontend already have a `User` type? `HasBetaFlags` barely provides any type safety as the only property it contains is optional.

Reply:
not yet, will have once I refactor auth controller in the client

Reply:
Hang on, isn't there an `IUser` type?